[이지예] 2주차 자동차 경주 게임 구현

jiyeeeah opened the pull request:
## 🚗 자동차 경주 기능 구현 목록

### 📌 게임 시작

- [x] 경주할 자동차의 이름 입력 받기 (쉼표를 기준으로 구분)
  - [x] 이름은 영문 5자 이하만 가능
- [x] 몇번 이동을 시도할 것인지 입력 받기
  - [x] 숫자만 입력 가능

### 📌 게임 진행

- [x] 게임 시작 문구 출력
- [x] 각 자동차마다 `0`에서 `9` 사이에서 무작위 값을 구한 후 값이 `4` 이상일 경우 전진 (입력받은 이동 횟수만큼 게임 진행)
  - [x] 전진하는 자동차를 출력할 때 자동차 이름을 같이 출력

### 📌 게임 완료

- [x] 최종 우승자 출력 (공동 우승자도 가능)

### 📌 프로그래밍 요구 사항

- [x] `indent`(인덴트, 들여쓰기) `depth`를 `3`이 넘지 않도록 구현.
- [x] `Jest`를 이용하여 본인이 정리한 기능 목록이 정상 동작함을 테스트 코드로 확인

---
## 🚗 자동차 경주 코드 구성

### `App.js`
- 게임이 전체적으로 진행되는 곳입니다.
- `게임 시작`, `게임 진행`, `게임 종료`로 나뉩니다.
- `게임 시작`
  - 자동차 이름을 입력받습니다.
  - 입력받은 자동차 이름들은 유효성 검사를 거쳐서 `Map`객체에 `자동차 이름 : 0`의 형태로 담깁니다.
- `게임 진행`
  - 입력 받은 시도 횟수 만큼 게임을 진행합니다.
  - 게임이 진행될 때마다 자동차들은 전진하거나 아무일도 일어나지 않습니다.
  - 시도 회차마다 결과를 출력합니다.
- `게임 종료`
  - 게임의 승자를 구하고 출력합니다.

### `utils`
- `move.js`
  - **자동차의 이동**을 위한 함수입니다.
  - 랜덤한 수를 구하고, 구한 값이 4 이상인 경우에만 전진합니다.
- `printResult.js`
  - 시도 회차가 끝날 때마다 **자동차들의 현 전진 상황을 출력**합니다.
- `validation.js`
  - 입력 값(자동차 이름, 시도한 횟수)가 **유효한지 검사**합니다.

---
## 🚗 소감문
  이전 과제([1주차 숫자 야구 게임](https://github.com/wooacourse-7th-study/javascript-baseball-6/pull/2)) 에서는 `Class`를 사용해서 그런지 조금 어렵다고 생각했는데, 단순 **함수 단위**로 쪼개서 구현해보니 훨씬 간단하게 느껴졌다. 이번 과제에서는 `commit` 메시지와 단위에 신경을 많이 썼다. 최대한 불필요한 `commit`이 없도록 `git rebase`를 적극 활용하였다. 추가 프로그래밍 요구 사항 중 `commit` 외에도 `Jest`를 활용한 테스트 코드 작성도 있었는데 이 부분에서 아쉽다. `Jest`를 조금 더 공부해보고 테스트 코드를 추가해보고 싶다.
  또한 지금 `move.js`와 `printResult.js`가 `utils` 디렉토리에 담겨 있는데, 이들이 `util`의 역할을 하는게 맞는지 고민이 된다. 보통 `util`이라 하면 전역적으로 여러번 사용될 때 재사용성을 높이기 위해 존재하는 것으로 아는데 이 두 함수는 한번씩 밖에 쓰이지 않기 때문이다. 물론 프로젝트의 규모가 작은 것도 있겠지만, 지난번 문제에서는 게임의 확장성까지 고려하여 util을 만들었는데, 이번에는 들여쓰기 조건을 맞추기 위한 util 함수들로 보이는 것이 아닌가 하는 생각이 든다. 리팩토링 과정에서 **게임의 확장성** 측면을 더 고려해봐야겠다. (이거 소감문 맞나..)
  이번 과제를 통해서 테스트 코드 작성법과 `commit`을 꼼꼼하게 관리하는 연습을 할 수 있었다. 앞으로도 적극적으로 활용해보면 도움이 많이 될 것 같다.

jiyeeeah committed:
[docs] 기능 구현 목록 작성

jiyeeeah committed:
[feat] 게임에 필요한 문구들 상수화

- 게임 진행에 필요한 입력, 출력 문구들 상수화하여 constants/message.js에 저장

jiyeeeah committed:
[feat] 경주할 자동차의 이름 입력 받기

- 게임 시작시, 경주할 자동차의 이름을 입력 받음
- 자동차의 이름은 쉼표로 구분하여 playersMap에 저장
- playersMap을 map객체로 하여 추후 전진 상황을 저장할 수 있도록 함
- 따라서 모든 player들은 value로 0을 갖고 시작함

jiyeeeah committed:
[feat] 자동차 이름 입력값 유효 체크

- 게임 시작 시, 경주할 자동차의 입력값에 대하여 5자 이하인지 체크
- 5자를 초과할 경우, false를 return
- 5자 이하인 경우, 입력받은 자동차들을 담은 배열 return

jiyeeeah committed:
[feat] 이동 횟수 입력 받는 기능

- 게임 시작시 전진 시도 횟수를 입력받음

jiyeeeah committed:
[feat] 시도 횟수 유효성 검사

- 게임 진행 횟수를 입력받을 때, 숫자 외의 값이 들어오지 않았는지 검사함

jiyeeeah committed:
[feat] 게임 시작 문구 출력

- 게임 시작을 알리는 "실행 결과" 문구를 출력함

jiyeeeah committed:
[feat] 자동차 전진 기능 구현

- 자동차 전진을 위한 move 함수 구현
- 자동차 Map을 돌며, 무작위 수를 구하고 구한 값이 4 이상인 경우만 value값이 +1이 되도록 함

jiyeeeah committed:
[feat] 게임 진행에 따라 전진 여부 출력

- 자동차의 전진 여부를 표시하는 printResult함수 구현

jiyeeeah committed:
[feat] 게임 종료시 회종 우승자 출력

- playersMap의 value값 중 가장 큰 값을 구하고,
  그 값을 가진 key를 winner 배열에 넣어줌

jiyeeeah committed:
[refactor] 유효성 검사 시 즉시 Error 발생시키도록 수정

- 유효성 검사 시에 throw 문으로 error 발생시키도록 함
- 상수에 ERROR_MESSAGES 추가하여 에러 메시지 관리

jiyeeeah committed:
[feat] 사용자 입력 값의 유효성 검사를 위한 test 코드 작성

- 사용자 입력 값 (자동차 이름, 시도 횟수)에 대하여 유효성 검사 코드가
  정상적으로 동작하는지 확인하기 위해 test 코드 작성함

jiyeeeah committed:
[feat] 게임 전체 테스트 코드 작성

- 게임 전체의 흐름을 테스트할 수 있는 코드 GameTest.js에 작성

jiyeeeah committed:
[docs] utils의 함수들 주석 작성

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,34 @@
-  async play() {}
+  async play() {
+    // 게임 시작
+    const playerInput = await MissionUtils.Console.readLineAsync(MESSAGES.START_INPUT);
+    const players = isPlayerInputValid(playerInput);
```
```suggestion
    const playerInput = await MissionUtils.Console.readLineAsync(MESSAGES.START_INPUT);
    const players = playerInput.split(",").map((v) => v.trim());
    playerInputValid(players);
```
`isPlayerInputValid` 함수 이름만 봤을 때는 인풋 값을 유효성 검사해서 `boolean` 값을 리턴할 것 같은데 막상 리턴하는 값은 유저 배열이기 때문에 조금 어색한 느낌이 있는 것 같아요!

`isPlayerInputValid` 함수안에 있는 `const players = input.split(",").map((v) => v.trim());` 이 로직을 밖을 빼고 `players`를 `isPlayerInputValid` 인자에 넣는건 어떨까요?? 

그리고 내부에서 `boolean` 값을 리턴하는 방식을 사용하거나 `isPlayerInputValid`의 함수 이름을 `playerInputValid`로 수정해도 좋을 것 같아요!

jiyeeeah replied:
오! 이게 훨씬 함수의 기능을 정확하게 할 수 있겠네요! 감사합니다 👍

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,34 @@
+    const playersMap = new Map(players.map((player) => [player, 0]));
+
+    const tryInput = await MissionUtils.Console.readLineAsync(MESSAGES.TRY_INPUT);
+    const tryNum = isTrialInputValid(tryInput);
```
```suggestion
    const tryNum = isTrialInputValid(tryInput);
```
`isTrialInputValid` 함수도 `boolean` 값을 반환하지 않는데 `is`라는 접두사를 사용해서 어색한 느낌이 있는 것 같아요🤔

jiyeeeah replied:
하나의 함수에서 여러가지를 할 수 있도록 하고 싶었는데 욕심이었던 것 같네요! 함수의 기능은 명확할수록 좋은건데..ㅠ 리뷰 감사합니당

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,34 @@
+    // 게임 진행
+    MissionUtils.Console.print(MESSAGES.RESULT);
+    for (let i = 0; i < tryNum; i++) {
+      move(playersMap);
```
```suggestion
      carsMove(playersMap);
```
`move` 라는 이름이 약간 함수 이름보다는 정적인 변수에 어울리는 것 같아요 조금 더 목적성을 드러내도 좋을 것 같아요!

wo-o29 commented on src/utils/printResult.js:
```diff
@@ -0,0 +1,15 @@
+ * @param {Map} playersMap
+ */
+export default function printResult(playersMap) {
+  playersMap.forEach((val, key) => {
+    let length = "";
+    for (let i = 0; i < val; i++) length += "-";
+    MissionUtils.Console.print(`${key} : ${length}`);
+  });
+  MissionUtils.Console.print("");
```
```suggestion
  playersMap.forEach((val, key) => {
    const dash = "-".repeat(val)
    MissionUtils.Console.print(`${key} : ${dash}`);
  });
  MissionUtils.Console.print("\n");
```
forEach 안에 내부 반복문 부분을 `repeat` 함수로 간략하게 수정할 수 있을 것 같아요!

jiyeeeah replied:
와 `repeat`...를 몰랐네요 ㅎㅎ 뭔가 바닐라 JS가 더 어려운 느낌...ㅋㅋㅋ 감사합니당

wo-o29 commented on src/utils/move.js:
```diff
@@ -0,0 +1,14 @@
+ * @param {Map} playersMap
+ */
+export default function move(playersMap) {
+  playersMap.forEach((val, key) => {
```
```suggestion
  playersMap.forEach((count, carName) => {
```
`val, key`도 괜찮은 것 같은데 조금 더 명확하게 표현해도 좋을 것 같아요!

jiyeeeah replied:
오오! 그러게요! 감사합니당

wo-o29 commented on src/App.js:
```diff
@@ -1,5 +1,34 @@
+    }
+
+    // 게임 끝
+    const maxLength = Math.max(...Array.from(playersMap.values()));
+    const winner = [];
+    playersMap.forEach((value, key) => {
+      if (value === maxLength) winner.push(key);
+    });
```
```suggestion
    const playersArray = [...playersMap.values()];
    const maxLength = Math.max(...playersArray);
    const winner = playersArray.filter((length) => length === maxLength);
```
여기 부분을 `filter`를 사용해서 간략하게 표현할 수 있을 것 같아요~!

jiyeeeah replied:
오오! 아직 저는 메서드에 대한 활용이 부족한 것 같아요ㅠ 감사합니다!

jiyeeeah replied:
음... 근데 제안해주신대로 수정하면 `playersArray`에는 `count`들만 들어있어서 각 자동차의 전진한 횟수만 들어있지 않을까요? 그럼 `winner`에는 최댓값 숫자만 남게 될 것 같아요..! 
저는 `Map`을 사용했기 때문에 `winner`를 얻기 위해서는 `value`와 `key`모두 필요해서 `forEach`를 사용했어요!

wo-o29 replied:
아 저기 playersArray가 들어갔군요 그러면 말씀해주신대로 playersMap을 fillter하시면 해결될 수 있을 것 같아요!

wo-o29 reviewed (approved):
저번부터 `Map` 같은 자료 구조를 잘 사용하시네요! 고수는 다르군요.. 많이 배웠습니다🔥

고생 많으셨습니다!👍🎯🙇‍♂️💪

ssilver01 commented on src/constants/message.js:
```diff
@@ -0,0 +1,13 @@
+
+// 게임 내 에러 메시지
+export const ERROR_MESSAGES = {
+  LIMIT: "[ERROR] 자동차 이름은 5자 이하만 가능합니다.",
+  NOT_NUMBER: "[ERROR] 숫자만 입력 가능합니다.",
```
LIMIT보단 CAR_LENGTH_LIMI과 같이 정확히 명시해 주면 더 좋을 것 같아요 !

ssilver01 commented on src/utils/validation.js:
```diff
@@ -0,0 +1,26 @@
+import { ERROR_MESSAGES } from "../constants/message.js";
+
+/**
+ * 자동차 이름을 입력했을 때 유효한지 검사합니다.
+ * @param {string} input
+ * @returns
+ */
```
각각 함수에 대해 param 설명과 함수 설명이 들어간 점 너무 좋은 것 같아요 !👍
```suggestion
/**
 * 자동차 이름을 입력했을 때 유효한지 검사합니다.
 * @param {string} input
 * @returns
 */
```

ssilver01 commented on src/utils/move.js:
```diff
@@ -0,0 +1,14 @@
+ */
+export default function move(playersMap) {
+  playersMap.forEach((val, key) => {
+    const randomNum = MissionUtils.Random.pickNumberInRange(0, 9);
+    if (randomNum < 4) return;
```
랜덤 숫자나 전진 가능한 숫자 범위를 상수화 해도 좋을 것 같아요

jiyeeeah replied:
이번엔 제가 게임의 확장성을 고려하지 못한 것 같아요ㅠㅠ 리뷰 감사드립니다! 👍

ssilver01 commented on src/App.js:
```diff
@@ -1,5 +1,34 @@
+    // 게임 끝
+    const maxLength = Math.max(...Array.from(playersMap.values()));
+    const winner = [];
+    playersMap.forEach((value, key) => {
```
value와 key 보단 각각의 의미를 잘 들어낼 수 있는 값을 쓰면 더 좋을 것 같아요

ssilver01 reviewed (commented):
테스트 코드도 게임진행, input값의 유효성 테스트에 따라 나눈 점 잘 한 것 같습니다. 
많이 배워가요 👍 수고하셨습니다⭐️

teagu123 commented on src/App.js:
```diff
@@ -1,5 +1,34 @@
+    }
+
+    // 게임 끝
+    const maxLength = Math.max(...Array.from(playersMap.values()));
+    const winner = [];
+    playersMap.forEach((value, key) => {
+      if (value === maxLength) winner.push(key);
+    });
```
App.js에서 보면 다른 로직들은 utils에 따로 함수를 만들어서 처리하는 모습인데 우승자를 만드는 로직만 따로 분리를 안해서 혹시 분리를 따로 안하시는 이유를 알수있을까요? 지금 utils들이 함수 하나당 하나의 역할만 수행하다보니 우승자를 정하는 함수도 만드는것도 좋아보입니다!!!

jiyeeeah replied:
소감문에도 적었지만 util함수들이 전부 Indent를 줄이기 위한 함수들 밖에 없는 것 같아서 그 점이 아쉬웠던 참이었어요.. ㅠㅠ 우승자 로직도 따로 분리하면 util 함수들의 역할이 명확해지니 좋을 것 같네요! 고민해보겠습니다. :) 감사합니다 👍

teagu123 commented on src/utils/move.js:
```diff
@@ -0,0 +1,14 @@
+    if (randomNum < 4) return;
+    playersMap.set(key, val + 1);
+  });
+}
```
저는 이걸 boolean으로 반환하게 진행하였는데 이런 방법도 좋네요! 많이 배웠습니다 👍

teagu123 commented on src/utils/validation.js:
```diff
@@ -0,0 +1,26 @@
+ * @param {string} input
+ * @returns
+ */
+export function isPlayerInputValid(input) {
+  const players = input.split(",").map((v) => v.trim());
+
+  for (let player of players) {
+    if (player.length > 5) throw new Error(ERROR_MESSAGES.LIMIT);
+  }
+
+  return players;
+}
```
입력의 빈값 예외처리가 보이지않아서 추가하셔야할거같습니다!!

jiyeeeah replied:
예외 처리가 아직 많이 부족하군요.. 감사합니다!

teagu123 commented on src/utils/validation.js:
```diff
@@ -0,0 +1,26 @@
+export function isTrialInputValid(input) {
+  if (isNaN(input)) throw new Error(ERROR_MESSAGES.NOT_NUMBER);
+  return Number(input);
+}
```
이것또한 빈값 예외처리와 1미만의 값이 들어올때는 예외처리를 진행해야할듯합니다!

jiyeeeah committed:
[refactor] move -> carsMove

- move 함수 네이밍 carsMove로 변경
- 자동차의 전진을 담당하는 함수임을 명시

jiyeeeah committed:
[refactor] forEach의 val, key -> count, player

- val, key가 각각 어떤걸 의미하는지 명시하기 위해 count와 player로 변경

jiyeeeah committed:
[refactor] repeat()사용 코드 간결화

- for문으로 dash의 길이를 정하고 print했었는데
  repeat()을 활용하여 dash를 만들어 코드를 간결하게 함

jiyeeeah committed:
[refactor] validation -> userInput

- validation파일의 유효성 검사를 userInput파일로 변경함
- 변경 후, 이 함수들 내에서 입력값을 받고 유효성 검사를 거쳐,
  App.js에서 사용될 수 있는 형태로 return하도록 함.

jiyeeeah committed:
[refactor] 예외 경우 추가

- 입력값이 공백일 경우, 시도할 횟수가 1 미만인 경우 등 추가
- 이에 적합한 에러메시지도 추가

jiyeeeah committed:
[feat] 추가된 유효성 검사 관련 test 추가

- 입력값 관련하여 추가된 유효성 검사에 대하여
  InputValidationTest.js에 test를 추가함
- 시도 횟수 입력값이 1 미만인 경우는 node src/index.js로 실행시,
  정상적으로 1 미만의 값은 입력할 수 없다는 메세지가 나오지만
  test 코드를 실행하면 숫자만 입력가능하다는 메세지가 나옴 => test를 통과 못함
  - 이 부분은 원인 파악 후 수정 예정. 현재 주석처리해둠

jiyeeeah committed:
[bugfix] test코드에서 NO_UNDER_1 에러메시지 안뜨던 오류 수정

- mockQuestions에서 shift()메서드를 제거하고,
  ERROR_CASE 배열의 input 부분을 배열이 아닌 문자열 형태로 수정함
